src/libsyntax/ext/deriving/generic/mod.rs

```diff
@@ -330,6 +330,39 @@ pub fn combine_substructure<'a>(f: CombineSubstructureFunc<'a>)
     RefCell::new(f)
 }
 
+fn is_generic(ty: &ast::Ty, ty_param_names: &[ast::Name]) -> bool {
+    use visit;
+
+    struct Visitor<'a> {
+        found: bool,
+        ty_param_names: &'a [ast::Name],
+    }
+
+    impl<'a> visit::Visitor<'a> for Visitor<'a> {
+        fn visit_path(&mut self, path: &'a ast::Path, _id: ast::NodeId) {
+            match path.segments.first() {
```

Comment:
I suspect this should check that `path.segments` has length 1, or maybe the shadowing rules means it doesn't matter? (i.e. is `struct Foo<std> { x: std::option::Option<u8> }` legal?)